### ISSUE TITLE
how to save pruned model?

### ISSUE THREAD

**Issue by cvJie** [BayesWatch/pytorch-prunes#1]
As question,
i saved the pruned model ,but model size no change.

**Comment by jack-willturner** [comment]
When you call `python train.py --deploy --eval <-other-commands->` the network is pruned at runtime (by the `if` statement on line 73) and then used.  

If you would like to save the pruned model, you could put a:
```
save_checkpoint({
                'epoch': SD['epoch'],
                'state_dict': model.state_dict(),
                'error_history': SD['error_history'],
            }, filename=filename)
```

after the call to `pruner.compress(model)` on line 89. 

However, it is better to leave the saved checkpoint in its original state, with the masks attached, to make it easier to recover the pruning trajectories. You can check the number of ops/params with `get_inf_params(model)` if you want an indication of compression rate. Whenever you want to use a pruned model for something, you then just need to run the code from lines 71 to 89 in `train.py`, followed by the task you want the pruned model to complete. 

Hope this helps.

**Comment by cvJie** [comment]
why not save pruned model in prune.py ?
and call train.py can to prune model ?
Now ,i want to use the pruned model to accelerate operation and reduce  model size..
can you give me some advices?

**Comment by jack-willturner** [comment]
Our reason for not saving pruned models in `prune.py` was to do with the objectives of the paper that this code accompanies (specifically, we needed `prune.py` to save the full pruning trajectory so that we could reinitialise them to train them from scratch). 

Calling `python train.py` with the flags `--deploy` and `--eval` does what you are asking. `--deploy` reduces the model size (although if you want to save it, you will need to copy and paste the snippet I described above), and `--eval` evaluates the newly reduced model on the CIFAR test set.

**Comment by cvJie** [comment]
Thanks

**Comment by cvJie** [comment]
Thanks
